### ISSUE TITLE
macho_handle_unixthread OOB

### ISSUE THREAD

**Issue by revskills** [VirusTotal/yara#1498]
After taking a look at the advisory [X41 D-Sec GmbH Security Advisory: X41-2021-001 - Multiple Vulnerabilities in YARA](https://www.x41-dsec.de/lab/advisories/x41-2021-001-yara/
) and fixes, just in case it is not reported already.

reachable in yr_rules_scan_mem(). And in order to reproduce it (tested on master) :

```
$ clang++ -fsanitize=address -std=c++11 -Ilibyara/include tests/oss-fuzz/macho_fuzzer.cc -o macho_test ./libyara/.libs/libyara.a
$ ./macho_test macho_handle_unixthread.bin

```

```
==21655==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6060000001ac at pc 0x00000055f952 bp 0x7fff5e043a30 sp 0x7fff5e043a28
READ of size 4 at 0x6060000001ac thread T0
    #0 0x55f951 in macho_handle_unixthread /vm/yara/libyara/modules/macho/macho.c:267:58
    #1 0x563079 in macho_parse_file /vm/yara/libyara/modules/macho/macho.c:624:9
    #2 0x569bef in macho__load /vm/yara/libyara/modules/macho/macho.c:1337:7
    #3 0x51d170 in yr_modules_load /vm/yara/libyara/modules.c:158:16
    #4 0x59b051 in yr_execute_code /vm/yara/libyara/exec.c:1444:16
    #5 0x526bd9 in yr_scanner_scan_mem_blocks /vm/yara/libyara/scanner.c:500:3
    #6 0x5277d6 in yr_scanner_scan_mem /vm/yara/libyara/scanner.c:644:16
    #7 0x523833 in yr_rules_scan_mem /vm/yara/libyara/rules.c:223:12
    #8 0x509f28 in LLVMFuzzerTestOneInput (/vm/yara/macho_test+0x509f28)
    #9 0x43d74b in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/vm/yara/macho_test+0x43d74b)
    #10 0x42604b in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/vm/yara/macho_test+0x42604b)
    #11 0x42c107 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/vm/yara/macho_test+0x42c107)
    #12 0x45a222 in main (/vm/yara/macho_test+0x45a222)
    #13 0x7fde031c9eba in __libc_start_main (/lib64/libc.so.6+0x23eba)
    #14 0x41fa19 in _start (/vm/yara/macho_test+0x41fa19)

Address 0x6060000001ac is a wild pointer.
SUMMARY: AddressSanitizer: heap-buffer-overflow /vm/yara/libyara/modules/macho/macho.c:267:58 in macho_handle_unixthread
Shadow bytes around the buggy address:
  0x0c0c7fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0c7fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0c7fff8000: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c0c7fff8010: 00 00 00 00 00 00 00 00 fa fa fa fa 00 00 00 00
  0x0c0c7fff8020: 00 00 04 fa fa fa fa fa 00 00 00 00 00 00 04 fa
=>0x0c0c7fff8030: fa fa fa fa fa[fa]fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8060: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8070: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8080: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==21655==ABORTING

```
[macho_handle_unixthread.bin.gz](https://github.com/VirusTotal/yara/files/6504372/macho_handle_unixthread.bin.gz)

**Comment by revskills** [comment]
after commit:
a11a7e5afe8957ba3be0f8af298c6149e21af1dc

```
=================================================================
==11869==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60600000026c at pc 0x00000055f95a bp 0x7ffc03277090 sp 0x7ffc03277088
READ of size 4 at 0x60600000026c thread T0
    #0 0x55f959 in macho_handle_unixthread /vm/yara/libyara/modules/macho/macho.c:272:58
    #1 0x563089 in macho_parse_file /vm/yara/libyara/modules/macho/macho.c:629:9
    #2 0x569bff in macho__load /vm/yara/libyara/modules/macho/macho.c:1342:7
    #3 0x51d170 in yr_modules_load /vm/yara/libyara/modules.c:158:16
    #4 0x59b061 in yr_execute_code /vm/yara/libyara/exec.c:1444:16
    #5 0x526bd9 in yr_scanner_scan_mem_blocks /vm/yara/libyara/scanner.c:500:3
    #6 0x5277d6 in yr_scanner_scan_mem /vm/yara/libyara/scanner.c:644:16
    #7 0x523833 in yr_rules_scan_mem /vm/yara/libyara/rules.c:223:12
    #8 0x509f28 in LLVMFuzzerTestOneInput (/vm/yara/macho_test+0x509f28)
    #9 0x43d74b in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/vm/yara/macho_test+0x43d74b)
    #10 0x42604b in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/vm/yara/macho_test+0x42604b)
    #11 0x42c107 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/vm/yara/macho_test+0x42c107)
    #12 0x45a222 in main (/vm/yara/macho_test+0x45a222)
    #13 0x7fe97fbe7eba in __libc_start_main (/lib64/libc.so.6+0x23eba)
    #14 0x41fa19 in _start (/vm/yara/macho_test+0x41fa19)

Address 0x60600000026c is a wild pointer.
SUMMARY: AddressSanitizer: heap-buffer-overflow /vm/yara/libyara/modules/macho/macho.c:272:58 in macho_handle_unixthread
Shadow bytes around the buggy address:
  0x0c0c7fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0c7fff8000: fa fa fa fa fd fd fd fd fd fd fd fa fa fa fa fa
  0x0c0c7fff8010: 00 00 00 00 00 00 05 fa fa fa fa fa fd fd fd fd
  0x0c0c7fff8020: fd fd fd fd fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c0c7fff8030: fa fa fa fa 00 00 00 00 00 00 04 fa fa fa fa fa
=>0x0c0c7fff8040: 00 00 00 00 00 00 04 fa fa fa fa fa fa[fa]fa fa
  0x0c0c7fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8060: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8070: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8080: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8090: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:	   fa
  Freed heap region:	   fd
  Stack left redzone:	   f1
  Stack mid redzone:	   f2
  Stack right redzone:     f3
  Stack after return:	   f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:	   f6
  Poisoned by user:        f7
  Container overflow:	   fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==11869==ABORTING
```

**Comment by plusvic** [comment]
Fixed in e87331beefe63df068481e4919b6007e6a05fbef.
